jlm/hls/opt/InvariantLambdaMemoryStateRemoval.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2025 Magnus Själander <work@sjalander.com>
+ * See COPYING for terms of redistribution.
+ */
+
+#include <jlm/hls/opt/InvariantLambdaMemoryStateRemoval.hpp>
+#include <jlm/llvm/ir/operators/MemoryStateOperations.hpp>
+#include <jlm/llvm/ir/operators/Phi.hpp>
+#include <jlm/rvsdg/lambda.hpp>
+#include <jlm/rvsdg/RvsdgModule.hpp>
+#include <jlm/rvsdg/traverser.hpp>
+#include <jlm/util/Statistics.hpp>
+
+namespace jlm::hls
+{
+
+void
+InvariantLambdaMemoryStateRemoval::RemoveInvariantMemoryStateEdges(
+    rvsdg::RegionResult * memoryState)
+{
+  // We only apply this for memory state edges that is invariant between LambdaEntryMemoryStateSplit
+  // and LambdaExitMemoryStateMerge nodes.
+  auto exitNode = rvsdg::output::GetNode(*memoryState->origin());
+  if (!rvsdg::is<const llvm::LambdaExitMemoryStateMergeOperation>(exitNode->GetOperation()))
+  {
+    return;
+  }
+
+  // Check if we have any invariant edge(s) between the two nodes
+  std::vector<size_t> indexes;
+  std::vector<rvsdg::output *> outputs;
+  rvsdg::Node * entryNode = nullptr;
+  for (size_t i = 0; i < exitNode->ninputs(); i++)
+  {
+    // Check if the output has only one user and if it is a LambdaEntryMemoryStateMerge
+    if (exitNode->input(i)->origin()->nusers() == 1)
+    {
+      auto node = rvsdg::output::GetNode(*exitNode->input(i)->origin());
+      if (jlm::rvsdg::is<const llvm::LambdaEntryMemoryStateSplitOperation>(node->GetOperation()))
+      {
+        // Found an invariant memory state edge, so going to replace the entryNode
+        entryNode = node;
+        continue;
+      }
+    }
+    // Keep track of edges that is to be kept since they are not invariant
+    outputs.push_back(exitNode->input(i)->origin());
+    // Also keep track of the index to be used for diverting edges
+    indexes.push_back(i);
+  }
+
+  // If the entryNode is not set, then we haven't found any invariant edges
+  if (entryNode == nullptr)
+  {
+    return;
+  }
+
+  // Replace LambdaEntryMemoryStateSplit and LambdaExitMemoryStateMerge nodes
+  if (outputs.size() == 0)
+  {
+    // The memory state edge is invariant, so we could in principle remove it from the lambda
+    // But the LLVM dialect expects to always have a memory state, so we connect the argument
+    // directly to the result
+    memoryState->divert_to(entryNode->input(0)->origin());
+  }
+  else if (outputs.size() == 1)
+  {
+    // Single edge that is not invariant, so we can elmintate the two MemoryState nodes
+    memoryState->divert_to(outputs.front());
+    entryNode->output(indexes.front())->divert_users(entryNode->input(0)->origin());
```

Comment:
This line should not be necessary, I believe. After the line above, the exit node and entry node should be dead.

Reply:
The first takes the output of the exitmerge node and connects it to the origin of the relevant input.
The second line takes the relevant output of the entrysplit node and connects it to the input.
I think both lines are needed to remove both nodes.

Reply:
You are absolutely right. It took me a while to fully understand these two lines.